### PR TITLE
TF-1842: handle task in worker manager

### PR DESCRIPTION
**Pull request by nqhhdev**
- issue: #1842 

- Resolved:

**`Foreground `**

https://github.com/linagora/tmail-flutter/assets/99852347/16b8352b-ac0d-427b-88e4-1d55c585e12a

`OIDC`

https://github.com/linagora/tmail-flutter/assets/99852347/f17b0584-5cd6-4399-a1cd-513f6d571f74

**`Background`**

https://github.com/linagora/tmail-flutter/assets/99852347/002ee80a-c001-4144-afb8-1d798c65d771

**`TERMINATED`**

https://github.com/linagora/tmail-flutter/assets/99852347/98e8abc1-4136-452a-a5e4-6f6fbd1cf92e

### CODE REVIEW

File: lib/features/composer/domain/model/sending_email.dart [hunk 1]
```diff
@@ -56,6 +56,28 @@ class SendingEmail with EquatableMixin {
     return val;
   }
 
+  factory SendingEmail.fromJson(Map<String, dynamic> json) {
+
+    return SendingEmail(
+      sendingId: json['sendingId'] as String,
+      email: Email.fromJson(jsonDecode(json['email'])),
+      emailActionType: _getEmailActionType(json['emailActionType'] as String),
+      sentMailboxId: json['sentMailboxId'] != null ? const MailboxIdNullableConverter().fromJson(json['sentMailboxId'] as String?) : null,
```
Comment (dab246) [inline lib/features/composer/domain/model/sending_email.dart:65]: You can write `sentMailboxId:  const MailboxIdNullableConverter().fromJson(json['sentMailboxId'] as String?)`

File: lib/features/composer/domain/model/sending_email.dart [hunk 2]
```diff
@@ -56,6 +56,28 @@ class SendingEmail with EquatableMixin {
...
+      emailIdDestroyed: json['emailIdDestroyed'] != null ? const EmailIdNullableConverter().fromJson(json['emailIdDestroyed'] as String?) : null,
```
Comment (dab246) [inline lib/features/composer/domain/model/sending_email.dart:66]: same

File: lib/features/composer/domain/model/sending_email.dart [hunk 3]
```diff
@@ -56,6 +56,28 @@ class SendingEmail with EquatableMixin {
...
+      emailIdAnsweredOrForwarded: json['emailIdAnsweredOrForwarded'] != null ? const EmailIdNullableConverter().fromJson(json['emailIdAnsweredOrForwarded'] as String?) : null,
```
Comment (dab246) [inline lib/features/composer/domain/model/sending_email.dart:67]: same

File: lib/features/composer/domain/model/sending_email.dart [hunk 4]
```diff
@@ -56,6 +56,28 @@ class SendingEmail with EquatableMixin {
...
+      identityId: json['identityId'] != null ? const IdentityIdNullableConverter().fromJson(json['identityId'] as String?) : null,
```
Comment (dab246) [inline lib/features/composer/domain/model/sending_email.dart:68]: same

File: lib/features/composer/domain/model/sending_email.dart [hunk 5]
```diff
@@ -56,6 +56,28 @@ class SendingEmail with EquatableMixin {
...
+      mailboxNameRequest: json['mailboxNameRequest'] != null ? MailboxName(json['mailboxNameRequest'] as String) : null,
```
Comment (dab246) [inline lib/features/composer/domain/model/sending_email.dart:69]: same

File: lib/features/composer/domain/model/sending_email.dart [hunk 6]
```diff
@@ -56,6 +56,28 @@ class SendingEmail with EquatableMixin {
...
+      creationIdRequest: json['creationIdRequest'] != null ? const IdNullableConverter().fromJson(json['creationIdRequest'] as String) : null,
```
Comment (dab246) [inline lib/features/composer/domain/model/sending_email.dart:70]: same

File: lib/features/composer/domain/model/sending_email.dart [hunk 7]
```diff
@@ -56,6 +56,28 @@ class SendingEmail with EquatableMixin {
...
+    );
+  }
+
+  static EmailActionType _getEmailActionType(String value) {
+    return EmailActionType.values.firstWhere(
+          (type) => type.toString().split('.').last == value,
```
Comment (dab246) [inline lib/features/composer/domain/model/sending_email.dart:76]:
```suggestion
(type) => type.name == value,
```

File: lib/features/email/data/datasource/email_datasource.dart [hunk 8]
```diff
@@ -91,4 +91,6 @@ abstract class EmailDataSource {
   Future<DetailedEmail?> getIncomingEmailedStored(Session session, AccountId accountId, EmailId emailId);
 
   Future<void> storeSendingEmail(AccountId accountId, UserName userName, SendingEmail sendingEmail);
+
+  Future<void> deleteSendingEmail(AccountId accountId, UserName userName, EmailId emailId);
```
Comment (dab246) [inline lib/features/email/data/datasource/email_datasource.dart:95]: Future<void> deleteSendingEmail(AccountId accountId, UserName userName, String sendingId);

File: lib/features/email/data/datasource_impl/email_datasource_impl.dart [hunk 9]
```diff
@@ -193,4 +193,9 @@ class EmailDataSourceImpl extends EmailDataSource {
   Future<void> storeSendingEmail(AccountId accountId, UserName userName, SendingEmail sendingEmail) {
     throw UnimplementedError();
   }
+
+  @override
+  Future<void> deleteSendingEmail(AccountId accountId, UserName userName, EmailId emailId) {
```
Comment (dab246) [inline lib/features/email/data/datasource_impl/email_datasource_impl.dart:198]: same

File: lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart [hunk 10]
```diff
@@ -251,4 +251,11 @@ class EmailHiveCacheDataSourceImpl extends EmailDataSource {
       return await _sendingEmailCacheManager.storeSendingEmail(accountId, userName, sendingEmail.toHiveCache());
     }).catchError(_exceptionThrower.throwException);
   }
+
+  @override
+  Future<void> deleteSendingEmail(AccountId accountId, UserName userName, EmailId emailId) {
```
Comment (dab246) [inline lib/features/email/data/datasource_impl/email_hive_cache_datasource_impl.dart:256]: same

File: lib/features/email/domain/repository/email_repository.dart [hunk 11]
```diff
@@ -103,4 +103,6 @@ abstract class EmailRepository {
   Future<DetailedEmail?> getOpenedEmail(Session session, AccountId accountId, EmailId emailId);
 
   Future<void> storeSendingEmail(AccountId accountId, UserName userName, SendingEmail sendingEmail);
+
+  Future<void> deleteSendingEmail(AccountId accountId, UserName userName, EmailId emailId);
```
Comment (dab246) [inline lib/features/email/domain/repository/email_repository.dart:107]: same

File: lib/features/email/data/repository/email_repository_impl.dart [hunk 12]
```diff
@@ -226,4 +226,9 @@ class EmailRepositoryImpl extends EmailRepository {
   Future<void> storeSendingEmail(AccountId accountId, UserName userName, SendingEmail sendingEmail) {
     return emailDataSource[DataSourceType.hiveCache]!.storeSendingEmail(accountId, userName, sendingEmail);
   }
+
+  @override
+  Future<void> deleteSendingEmail(AccountId accountId, UserName userName, EmailId emailId) {
```
Comment (dab246) [inline lib/features/email/data/repository/email_repository_impl.dart:231]: same

File: lib/features/email/domain/state/delete_sending_email_state.dart [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+
+class DeleteSendingEmailLoading extends UIState {}
+
+class DeleteSendingEmailSuccess extends UIState {
```
Comment (dab246) [inline lib/features/email/domain/state/delete_sending_email_state.dart:6]:
```suggestion
class DeleteSendingEmailSuccess extends UIState {}
```

File: lib/features/email/domain/usecases/delete_sending_email_interactor.dart [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+import 'package:jmap_dart_client/jmap/mail/email/email.dart';
+import 'package:tmail_ui_user/features/email/domain/repository/email_repository.dart';
+import 'package:tmail_ui_user/features/email/domain/state/delete_sending_email_state.dart';
+
+class DeleteSendingEmailInteractor {
+  final EmailRepository _emailRepository;
+
+  DeleteSendingEmailInteractor(this._emailRepository);
+
+  Stream<Either<Failure, Success>> execute(AccountId accountId, UserName userName, EmailId emailId) async* {
```
Comment (dab246) [inline lib/features/email/domain/usecases/delete_sending_email_interactor.dart:15]:
```suggestion
Stream<Either<Failure, Success>> execute(AccountId accountId, UserName userName, String sendingId) async* {
```

File: lib/features/offline_mode/manager/sending_email_cache_manager.dart [hunk 15]
```diff
@@ -22,4 +22,14 @@ class SendingEmailCacheManager {
     log('SendingEmailCacheManager::storeSendingEmail():keyCache: $keyCache | sendingEmailHiveCache: $sendingEmailHiveCache');
     return _hiveCacheClient.insertItem(keyCache, sendingEmailHiveCache);
   }
+
+  Future<void> deleteSendingEmail(
+    AccountId accountId,
+    UserName userName,
+    String sendingEmailId
```
Comment (dab246) [inline lib/features/offline_mode/manager/sending_email_cache_manager.dart:29]:
```suggestion
String sendingId
```

File: lib/features/offline_mode/controller/work_scheduler_controller.dart [hunk 16]
```diff
@@ -63,14 +64,15 @@ class WorkSchedulerController {
         final matchedType = WorkerType.values.firstWhereOrNull((type) => type.name == workerType);
 
         if (matchedType != null) {
-          await matchedType.usingObserver().bindDI();
-          await matchedType.usingObserver().observe(taskName, dataObject);
+          await matchedType.usingObserver().bindDI(completer);
+          await matchedType.usingObserver().observe(taskName, dataObject, completer);
         }
```
Comment (dab246) [inline lib/features/offline_mode/controller/work_scheduler_controller.dart:69]: Just pass `completer` to bindingDI

File: lib/features/offline_mode/observer/sending_email_observer.dart [hunk 17]
```diff
@@ -1,24 +1,253 @@
...
+      SendingEmailBindings().dependencies();
+    });
+
+    return await _getInteractorBindings();
+  }
+
+  @override
+  void handleFailureViewState(Failure failure) {
+    _clearDataQueue();
+    log('SendingEmailObserver::_handleFailureViewState(): $failure');
```
Comment (dab246) [inline lib/features/offline_mode/observer/sending_email_observer.dart:97]: should call `completer.completerError();`

File: lib/features/offline_mode/observer/sending_email_observer.dart [hunk 18]
```diff
@@ -1,24 +1,253 @@
...
+      _getSessionInteractor = getBinding<GetSessionInteractor>();
+      _appToast = getBinding<AppToast>();
+      _imagePaths = getBinding<ImagePaths>();
+      _sendEmailInteractor = getBinding<SendEmailInteractor>();
+      _deleteSendingEmailInteractor = getBinding<DeleteSendingEmailInteractor>();
+    } catch (e) {
+      logError('SendingEmailObserver::_getInteractorBindings(): ${e.toString()}');
+      return _completer?.completeError(false);
+    }
+    return _completer?.complete(true);
```
Comment (dab246) [inline lib/features/offline_mode/observer/sending_email_observer.dart:132]: why call here. If you do so you will terminate the workmanager

File: lib/features/offline_mode/observer/sending_email_observer.dart [hunk 19]
```diff
@@ -1,24 +1,253 @@
...
+
+    _dynamicUrlInterceptors?.changeBaseUrl(storedTokenOidcSuccess.baseUrl.toString());
+    _getSessionAction();
+  }
+
+  void _handleSendEmailSuccess(SendEmailSuccess success) async {
+    log("SendingEmailObserver::_handleSendEmailSuccess()");
+    _deleteSendingEmailAction();
+    if (currentOverlayContext != null && currentContext != null) {
+      _appToast?.showToastSuccessMessage(
```
Comment (dab246) [inline lib/features/offline_mode/observer/sending_email_observer.dart:224]: remove it. We only push notification when

File: lib/features/offline_mode/observer/sending_email_observer.dart [hunk 20]
```diff
@@ -1,24 +1,253 @@
...
+  }
+
+  void _handleDeleteSendingEmailSuccess() async {
+    log('SendingEmailObserver::_handleDeleteSendingEmailSuccess(): Success');
+    return _completer?.complete(true);
+  }
+
+  void _clearDataQueue() async {
+    _sendingEmail = null;
+    return _completer?.completeError(false);
```
Comment (dab246) [inline lib/features/offline_mode/observer/sending_email_observer.dart:251]:
```suggestion
_completer?.completeError(SendingEmailFromWorkmanagerException());
```

File: lib/features/base/base_controller.dart [hunk 21]
```diff
@@ -243,9 +243,6 @@ abstract class BaseController extends GetxController
         FcmConfiguration.initialize();
         FcmInteractorBindings().dependencies();
         FcmMessageController.instance.initializeFromAccountId(accountId, session);
-        if (!BuildUtils.isWeb) {
-          LocalNotificationManager.instance.setUp();
-        }
```
Comment (hoangdat) [inline lib/features/base/base_controller.dart:248]: Why remove it?

**Comment by hoangdat** [comment]
Please demo the cases: app in background/terminated/foreground when the network comeback

**Comment by hoangdat** [comment]
Please demo the case: 
1. App in offline
2. Sending email
3. Queue work well
4. Logout the app
5. Observe the behavior

File: lib/features/base/base_controller.dart (hunk 21 above)
Comment (nqhhdev) [reply lib/features/base/base_controller.dart:248]: I move it to `main()` for init with workmanager

File: lib/features/offline_mode/observer/sending_email_observer.dart [hunk 22]
```diff
@@ -1,24 +1,250 @@
...
+import 'package:tmail_ui_user/main/localizations/app_localizations.dart';
+import 'package:tmail_ui_user/main/routes/route_navigation.dart';
 
 class SendingEmailObserver extends WorkObserver {
 
+  AccountId? _currentAccountId;
+  Session? _currentSession;
+  UserName? _userName;
+  SendingEmail? _sendingEmail;
+  String? _sendingId;
```
Comment (dab246) [inline lib/features/offline_mode/observer/sending_email_observer.dart:43]: remove

File: lib/features/offline_mode/observer/sending_email_observer.dart [hunk 23]
```diff
@@ -1,24 +1,250 @@
...
   }
 
   @override
   Future<void> bindDI() async {
-    log('SendingEmailObserver::bindDI(): ');
-    return Future.value();
+    return await _initialAppConfig();
+  }
+
+
```
Comment (dab246) [inline lib/features/offline_mode/observer/sending_email_observer.dart:75]: remove line

File: lib/main.dart [hunk 24]
```diff
@@ -26,6 +27,7 @@ void main() async {
        Executor().warmUp(),
        if (PlatformInfo.isMobile)
         WorkManagerConfig().initialize(),
+        LocalNotificationManager.instance.setUp(),
```
Comment (dab246) [inline lib/main.dart:30]: only initial on mobile

**Review by dab246 — APPROVED** [review]

**Pull request merged**